File: Collector.java
package nstreamcom;

public class Collector {
    public enum States {

        COLLECTED,
        MISSING_SIZE,
        MISSING_DATA,
        WAITING_SIZE,
        WAITING_DATA;

        private boolean _errorState;

        static {
            COLLECTED._errorState = false;
            MISSING_SIZE._errorState = true;
            MISSING_DATA._errorState = true;
            WAITING_SIZE._errorState = false;
            WAITING_DATA._errorState = false;
        }

        public boolean error() {
            return _errorState;
        }
    }

    private States _state = States.WAITING_SIZE;

    private final BufferedDecoder _decoder = new BufferedDecoder();

    private int _nextSize;

    public States state() {
        return _state;
    }

    public int nextSize() {
        if (_state == States.COLLECTED || _state == States.WAITING_DATA) {
            return _nextSize;
        }
        return 0;
    }

    public byte[] bytes() {
        if (_state == States.COLLECTED) {
            return _decoder.bytes();
        }
        return new byte[0];
    }

    public void reset() {
        _nextSize = 0;
        _decoder.reset();
        _state = States.WAITING_SIZE;
    }

    public States collect(byte b) {
        switch (_state) {
            case COLLECTED:
            case MISSING_SIZE:
            case MISSING_DATA:
                reset();
            case WAITING_SIZE:
                collectSizeByte(b);
                break;
            case WAITING_DATA:
                collectDataByte(b);
                break;
            default:
                break;
        }
        return _state;
    }

    private void collectSizeByte(byte b) {
        if ((b & (1 << 7)) == 0) {
            _state = States.MISSING_SIZE;
            return;
        }

        b &= (byte)~(0xFF << NSize.DATA_BITS);
        boolean last = _decoder.decodedIndex() == NSize.NSIZE_SIZE;
        _decoder.next(b, last);

        if (!last) {
            return;
        }

        _nextSize = (int)NSize.littleUInt32toLong(_decoder.bytes());
        _decoder.reset();
        _state = _nextSize == 0 ? States.COLLECTED : States.WAITING_DATA;
    }

    private void collectDataByte(byte b) {
        if ((b & (1 << 7)) > 0) {
            _state = States.MISSING_SIZE;
            return;
        }

        boolean last = _decoder.decodedIndex() == _nextSize;
        _decoder.next(b, last);

        if (!last) {
            return;
        }

        _state = States.COLLECTED;
    }
}

File: src/collector.zig
const std = @import("std");
const sizing = @import("sizing.zig");
const buffered_decoder = @import("buffered_decoder.zig");

pub const CollectorState = enum {
    Collected,
    WaitingSize,
    WaitingData,
    MissingSize,
    MissingData,
    BufferFull
};

pub const CollectorError = error {
    MissingSize,
    MissingData
};

pub const Error = CollectorError || buffered_decoder.DecoderError;

pub const Collector = struct {
    state: CollectorState = CollectorState.WaitingSize,
    next_size: sizing.nsize_int = 0,
    decoder: *buffered_decoder.Decoder,

    pub fn sizeReady(self: *const Collector) bool {
        return self.state == CollectorState.Collected or self.state == Collector.WaitingData;
    }

    pub fn errorState(self: *const Collector) bool {
        return
            self.state == CollectorState.MissingSize or
            self.state == CollectorState.MissingData or
            self.state == CollectorState.BufferFull;
    }

    pub fn reset(self: *Collector) void {
        self.state = CollectorState.WaitingSize;
        self.next_size = 0;
        self.decoder.reset();
    }

    pub fn collect(self: *Collector, byte: u8) Error!void {
        switch (self.state) {
            CollectorState.MissingSize, CollectorState.MissingData => {
                self.reset();
                return self.collectSizeByte(byte);
            },
            CollectorState.Collected, CollectorState.WaitingSize => return self.collectSizeByte(byte),
            CollectorState.WaitingData => return self.collectDataByte(byte),
            else => unreachable,
        }
    }

    fn collectSizeByte(self: *Collector, byte: u8) Error!void {
        if ((byte & 1 << 7) == 0) {
            self.state = CollectorState.MissingSize;
            return Error.MissingSize;
        }

        const is_last = self.decoder.position == sizing.nsize_int_byte_count;
        try self.decoder.next(byte & ~(@as(u8, 0xFF) << 7), is_last);

        if (!is_last) {
            return;
        }

        self.next_size = std.mem.bytesAsValue(sizing.nsize_int, self.decoder.buffer.?).*;
        self.decoder.reset();
        self.state = if (self.next_size == 0) CollectorState.Collected else CollectorState.WaitingData;
    }

    fn collectDataByte(self: *Collector, byte: u8) Error!void {
        if ((byte & 1 << 7) > 0) {
            self.state = CollectorState.MissingData;
            return;
        }

        const last= self.decoder.position == self.next_size;
        try self.decoder.next(byte, last);

        if (!last) {
            return;
        }

        self.state = CollectorState.Collected;
    }
};

